rustls/src/client/builder.rs

```diff
@@ -161,6 +166,97 @@ impl ConfigBuilder<ClientConfig, WantsClientCert> {
             enable_early_data: false,
             #[cfg(feature = "tls12")]
             require_ems: cfg!(feature = "fips"),
+            time_provider: Arc::new(DefaultTimeProvider),
+        }
+    }
+}
+
+// When the std feature is not enabled we have an extra state in the config builder process
+// for providing a time provider.
+#[cfg(not(feature = "std"))]
+impl ConfigBuilder<ClientConfig, WantsClientCert> {
+    /// Sets a single certificate chain and matching private key for use
+    /// in client authentication.
+    ///
+    /// `cert_chain` is a vector of DER-encoded certificates.
+    /// `key_der` is a DER-encoded private key as PKCS#1, PKCS#8, or SEC1. The
+    /// `aws-lc-rs` and `ring` [`CryptoProvider`]s support all three encodings,
+    /// but other `CryptoProviders` may not.
+    ///
+    /// This function fails if `key_der` is invalid.
+    pub fn with_client_auth_cert(
+        self,
+        cert_chain: Vec<CertificateDer<'static>>,
+        key_der: PrivateKeyDer<'static>,
+    ) -> Result<ConfigBuilder<ClientConfig, WantsTimeProvider>, Error> {
+        let private_key = self
+            .state
+            .provider
+            .key_provider
+            .load_private_key(key_der)?;
+        let resolver =
+            handy::AlwaysResolvesClientCert::new(private_key, CertificateChain(cert_chain))?;
+        Ok(self.with_client_cert_resolver(Arc::new(resolver)))
+    }
+
+    /// Do not support client auth.
+    pub fn with_no_client_auth(self) -> ConfigBuilder<ClientConfig, WantsTimeProvider> {
+        self.with_client_cert_resolver(Arc::new(handy::FailResolveClientCert {}))
+    }
+
+    /// Sets a custom [`ResolvesClientCert`].
+    pub fn with_client_cert_resolver(
+        self,
+        client_auth_cert_resolver: Arc<dyn ResolvesClientCert>,
+    ) -> ConfigBuilder<ClientConfig, WantsTimeProvider> {
+        ConfigBuilder {
+            state: WantsTimeProvider {
+                provider: self.state.provider,
+                versions: self.state.versions,
+                verifier: self.state.verifier,
+                client_auth_cert_resolver,
+            },
+            side: PhantomData,
+        }
+    }
+}
+
+/// A config builder state where the caller needs to supply a [`TimeProvider`].
+///
+/// For more information, see the [`ConfigBuilder`] documentation.
+///
+/// [`TimeProvider`]: crate::time_provider::TimeProvider
+#[cfg(not(feature = "std"))]
+#[derive(Clone)]
+pub struct WantsTimeProvider {
+    provider: Arc<CryptoProvider>,
+    versions: versions::EnabledVersions,
+    verifier: Arc<dyn verify::ServerCertVerifier>,
+    client_auth_cert_resolver: Arc<dyn ResolvesClientCert>,
+}
+
+#[cfg(not(feature = "std"))]
```

Comment:
Would the appeal be maintaining symmetry, or is there a deeper usecase you had in mind where you would want to provide a time provider w/ `std` enabled?

I don't have a strong preference here so would generally favour whatever Djc thinks is best since his feedback prompted the initial change from the approach used by the PR's first author.

Reply:
> Sorry for the drive-by comment

Happy for input here btw :-)

Reply:
It feels less hazardous (in the sense of crate features changing return types, etc.) to:

- std-gate all existing `ClientConfig::builder*`
- add ungated `ClientConfig::builder_with_details` (or some better name) that takes absolutely everything: a CryptoProvider, this time provider, (potentially a `Resumption` to put the no-std unhelpful and surprising `Resumption::default` front-and-center?)

Reply:
Yeah, hiding API when a feature is enabled is generally a no-no because Cargo features are supposed to be additive.

Reply:
To clarify: do you mean adding a parameter to `builder_with_provider`? Or is a new `builder_with_*` function OK?

(I think my opinion is: 100% of `builder_with_provider` callers are std currently, and inflicting that change on them seems a bit gratuitous; so I prefer a new API surface.)

Reply:
I have pushed a commit (that can be dissolved into earlier ones) reflecting my preferences on this.

Reply:
Yup, that's pretty much exactly what I had in mind. Tried to explain (badly) in https://github.com/rustls/rustls/pull/1502#discussion_r1434942346.

Reply:
I've squashed this into the commits that introduced it. I also took the liberty of improving some of the docs around `TimeProvider`

Reply:
This direction makes sense to me :+1: